cli/cli.ts

```diff
@@ -1588,13 +1613,52 @@ function buildTargetCoreAsync() {
                 .map(p => path.join("sim", p))
                 .filter(p => fs.statSync(p).isDirectory()));
     }
+
+    let hexCachePath = path.resolve(process.cwd(), "built", "hexcache");
+    nodeutil.deleteFolderRecursive(hexCachePath);
+    nodeutil.mkdirP(hexCachePath);
+
     console.log(`building target.json in ${process.cwd()}...`)
-    return forEachBundledPkgAsync(pkg =>
-        pkg.filesToBePublishedAsync()
+    return forEachBundledPkgAndTemplateAsync((pkg, dirname) => {
+        pxt.log(`building in ${dirname}`);
+        let isTemplate = dirname.indexOf(path.join("libs", "templates")) === 0;
+
+        if (isTemplate) {
+            forceCloudBuild = true;
+        } else {
+            forceCloudBuild = previousForceCloudBuild;
+        }
+
+        return pkg.filesToBePublishedAsync(true)
             .then(res => {
                 cfg.bundledpkgs[pkg.config.name] = res
             })
-            .then(testForBuildTargetAsync))
+            .then(testForBuildTargetAsync)
+            .then((compileOpts) => {
+                // For the templates, we need to save the base HEX file to the offline HEX cache
+                if (isTemplate) {
+                    if (!compileOpts) {
+                        console.error(`Failed to extract HEX image for template ${dirname}`);
+                        return;
+                    }
+
+                    // Place the base HEX image in the hex cache if necessary
+                    let sha = compileOpts.extinfo.sha;
+                    let hex: string[] = compileOpts.hexinfo.hex;
+                    let hexFile = path.join(hexCachePath, sha + ".hex");
+
+                    if (fs.existsSync(hexFile)) {
+                        pxt.log(`HEX image already in offline cache for template ${dirname}`);
+                    } else {
+                        fs.writeFileSync(hexFile, hex.join(os.EOL));
```

Comment:
When the HEX info is returned, we are already intelligent about Windows vs Unix line endings, see
https://github.com/Microsoft/pxt/blob/master/pxtlib/cpp.ts#L832